File: go-client/protocol.go
package main

import (
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	// "bufio"
)

type PeerConn struct {
	Name       string
	Conn       interface{ Close() error }
	RW         interface {
		ReadString(byte) (string, error)
		WriteString(string) (int, error)
		Flush() error
	}
	Key        []byte
	RemotePub  []byte
	// RemotePub ed25519.PublicKey
	RemoteAddr string
}

type App struct {
	Self  *Identity
	Conns map[string]*PeerConn
	Mu    sync.Mutex
}

func NewApp(self *Identity) *App {
	return &App{
		Self:  self,
		Conns: make(map[string]*PeerConn),
	}
}

func (a *App) AddConn(pc *PeerConn) {
	a.Mu.Lock()
	defer a.Mu.Unlock()
	if _, exists := a.Conns[pc.Name]; exists {
		_ = pc.Conn.Close()
		return
	}
	a.Conns[pc.Name] = pc
}

func (a *App) GetConn(name string) (*PeerConn, bool) {
	a.Mu.Lock()
	defer a.Mu.Unlock()
	pc, ok := a.Conns[name]
	return pc, ok
}

func (a *App) ListPeers() []string {
	a.Mu.Lock()
	defer a.Mu.Unlock()
	out := make([]string, 0, len(a.Conns))
	for name := range a.Conns {
		out = append(out, name)
	}
	return out
}

func (a *App) HasPeer(addr string) bool {
    a.Mu.Lock()
    defer a.Mu.Unlock()

    for _, p := range a.Conns {
        if p.RemoteAddr == addr {
            return true
        }
    }
    return false
}

func sendEncrypted(pc *PeerConn, payload string) error {
	nonce, ciphertext, err := Encrypt(pc.Key, []byte(payload))
	if err != nil {
		return err
	}
	line := fmt.Sprintf(
		"DATA|%s|%s\n",
		base64.StdEncoding.EncodeToString(nonce),
		base64.StdEncoding.EncodeToString(ciphertext),
	)
	if _, err := pc.RW.WriteString(line); err != nil {
		return err
	}
	return pc.RW.Flush()
}

func listSharedFiles() []string {
	_ = os.MkdirAll("shared_files", 0755)
	entries, err := os.ReadDir("shared_files")
	if err != nil {
		return []string{}
	}
	out := []string{}
	for _, e := range entries {
		if !e.IsDir() {
			out = append(out, e.Name())
		}
	}
	return out
}

func saveDownload(filename string, data []byte) error {
	 _ = os.MkdirAll("downloads", 0755)
	path := filepath.Join("downloads", filepath.Base(filename))
	// return os.WriteFile(path, data, 0644)
	key := sha256Bytes([]byte("local-secret-key")) // simple local key

	nonce, ciphertext, err := Encrypt(key, data)
	if err != nil {
		return err
	}

	payload := base64.StdEncoding.EncodeToString(nonce) + "|" +
		base64.StdEncoding.EncodeToString(ciphertext)

	return os.WriteFile(path, []byte(payload), 0600)
	
}

func metadataPath(filename string) string {
	return filepath.Join("downloads", filepath.Base(filename)+".meta")
}

func sharedMetadataPath(filename string) string {
	return filepath.Join("shared_files", filepath.Base(filename)+".meta")
}

func saveMetadata(filename, originName string, originPub, hash, sig []byte) error {
	content := strings.Join([]string{
		originName,
		base64.StdEncoding.EncodeToString(originPub),
		base64.StdEncoding.EncodeToString(hash),
		base64.StdEncoding.EncodeToString(sig),
	}, "|")
	return os.WriteFile(metadataPath(filename), []byte(content), 0600)
}

func loadMetadata(path string) (originName string, originPub, hash, sig []byte, err error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", nil, nil, nil, err
	}
	parts := strings.Split(strings.TrimSpace(string(data)), "|")
	if len(parts) != 4 {
		return "", nil, nil, nil, fmt.Errorf("invalid metadata format")
	}

	originName = parts[0]
	originPub, err = base64.StdEncoding.DecodeString(parts[1])
	if err != nil {
		return "", nil, nil, nil, err
	}
	hash, err = base64.StdEncoding.DecodeString(parts[2])
	if err != nil {
		return "", nil, nil, nil, err
	}
	sig, err = base64.StdEncoding.DecodeString(parts[3])
	if err != nil {
		return "", nil, nil, nil, err
	}
	return originName, originPub, hash, sig, nil
}

func processPayload(a *App, pc *PeerConn, payload string) {
	parts := strings.Split(payload, "|")
	switch parts[0] {
	case "PING":
		fmt.Printf("[%s] PING received\n", pc.Name)

	case "LIST_REQ":
		files := strings.Join(listSharedFiles(), ",")
		_ = sendEncrypted(pc, "LIST_RES|"+files)

	case "LIST_RES":
		if len(parts) < 2 || parts[1] == "" {
			fmt.Printf("[%s] shared files: (none)\n", pc.Name)
			return
		}
		fmt.Printf("[%s] shared files: %s\n", pc.Name, parts[1])

	case "GET_REQ":
		if len(parts) < 2 {
			_ = sendEncrypted(pc, "ERROR|missing filename")
			return
		}
	
		filename := filepath.Base(parts[1])
	
		fmt.Printf("[%s] wants file '%s'. Accept? (y/n): ", pc.Name, filename)
	
		var resp string
		fmt.Scanln(&resp)
	
		if strings.ToLower(strings.TrimSpace(resp)) != "y" {
			_ = sendEncrypted(pc, "ERROR|request denied")
			return
		}
	
		var data []byte
		var hash []byte
		var sig []byte
		var originPub []byte
		var originName string
	
		sharedPath := filepath.Join("shared_files", filename)
		downloadPath := filepath.Join("downloads", filename)
	
		if _, err := os.Stat(sharedPath); err == nil {
			// File is locally owned or explicitly shared
			data, err = os.ReadFile(sharedPath)
			if err != nil {
				_ = sendEncrypted(pc, "ERROR|file not found")
				return
			}
	
			metaPath := sharedMetadataPath(filename)
			if _, err := os.Stat(metaPath); err == nil {
				originName, originPub, hash, sig, err = loadMetadata(metaPath)
				if err != nil {
					_ = sendEncrypted(pc, "ERROR|invalid metadata")
					return
				}
			} else {
				// This peer is the original owner
				hash = sha256Bytes(data)
				sig = signHash(a.Self.Priv, hash)
				originPub = a.Self.Pub
				originName = a.Self.Name
			}
		} else if _, err := os.Stat(downloadPath); err == nil {
			// File was downloaded earlier; serve with original metadata
			encPayload, err := os.ReadFile(downloadPath)
			if err != nil {
				_ = sendEncrypted(pc, "ERROR|file not found")
				return
			}
	
			encParts := strings.Split(string(encPayload), "|")
			if len(encParts) != 2 {
				_ = sendEncrypted(pc, "ERROR|invalid stored file")
				return
			}
	
			localKey := sha256Bytes([]byte("local-secret-key"))
			nonce, err1 := base64.StdEncoding.DecodeString(encParts[0])
			ciphertext, err2 := base64.StdEncoding.DecodeString(encParts[1])
			if err1 != nil || err2 != nil {
				_ = sendEncrypted(pc, "ERROR|invalid stored file")
				return
			}
	
			data, err = Decrypt(localKey, nonce, ciphertext)
			if err != nil {
				_ = sendEncrypted(pc, "ERROR|failed to decrypt stored file")
				return
			}
	
			originName, originPub, hash, sig, err = loadMetadata(metadataPath(filename))
			if err != nil {
				_ = sendEncrypted(pc, "ERROR|missing original metadata")
				return
			}
		} else {
			_ = sendEncrypted(pc, "ERROR|file not found")
			return
		}
	
		msg := "GET_RES|" +
			filename + "|" +
			base64.StdEncoding.EncodeToString(data) + "|" +
			base64.StdEncoding.EncodeToString(hash) + "|" +
			base64.StdEncoding.EncodeToString(sig) + "|" +
			base64.StdEncoding.EncodeToString(originPub) + "|" +
			originName
	
		_ = sendEncrypted(pc, msg)

	case "GET_RES":
		if len(parts) < 7 {
			fmt.Printf("[%s] malformed GET_RES\n", pc.Name)
			return
		}
		// filename := parts[1]
		filename := filepath.Base(parts[1])
		fileData, err1 := base64.StdEncoding.DecodeString(parts[2])
		hash, err2 := base64.StdEncoding.DecodeString(parts[3])
		sig, err3 := base64.StdEncoding.DecodeString(parts[4])
		originPub, err4 := base64.StdEncoding.DecodeString(parts[5])
		originName := parts[6]
		if err1 != nil || err2 != nil || err3 != nil || err4 != nil{
			fmt.Printf("[%s] failed to decode GET_RES\n", pc.Name)
			return
		}

		actualHash := sha256Bytes(fileData)
		if !equalBytes(actualHash, hash) {
			// fmt.Printf("[%s] hash mismatch for %s\n", pc.Name, filename)
			fmt.Printf("[%s] integrity check failed for %s\n", pc.Name, filename)
			return
		}
		if !verifyHash(originPub, hash, sig) {
			// fmt.Printf("[%s] signature verification failed for %s\n", pc.Name, filename)
			fmt.Printf("[%s] signature verification FAILED for %s\n", pc.Name, filename)
			return
		}
		if err := saveDownload(filename, fileData); err != nil {
			fmt.Printf("[%s] save failed: %v\n", pc.Name, err)
			return
		}
		
		if err := saveMetadata(filename, originName, originPub, hash, sig); err != nil {
			fmt.Printf("[%s] metadata save failed: %v\n", pc.Name, err)
			return
		}

		fmt.Printf("[%s] downloaded and verified %s (original owner: %s)\n", pc.Name, filename, originName)

	case "ERROR":
		if len(parts) >= 2 {
			fmt.Printf("[%s] ERROR: %s\n", pc.Name, parts[1])
		}
	}
}

File: java-client/Protocol.java
import java.io.*;
import java.nio.file.*;
import java.security.MessageDigest;
import java.util.*;

import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters;

import java.security.*;
import java.util.Base64;
import org.bouncycastle.crypto.params.*;
import org.bouncycastle.crypto.generators.X25519KeyPairGenerator;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.agreement.X25519Agreement;

class PeerConn {
    public final String name;
    public final java.net.Socket socket;
    public final BufferedReader reader;
    public final BufferedWriter writer;
    public final byte[] key;
    public byte[] remotePub;

    public PeerConn(String name, java.net.Socket socket, BufferedReader reader,
                    BufferedWriter writer, byte[] key, byte[] remotePub) {
        this.name = name;
        this.socket = socket;
        this.reader = reader;
        this.writer = writer;
        this.key = key;
        this.remotePub = remotePub;
    }
}

public class Protocol {
    public static void ensureDirs(String filename, String content) throws Exception {
        Files.createDirectories(Paths.get("shared_files"));
        Files.createDirectories(Paths.get("downloads"));
        Path p = Paths.get("shared_files", filename);
        if (!Files.exists(p)) {
            Files.writeString(p, content);
        }
    }

    public static List<String> listSharedFiles() throws Exception {
        List<String> out = new ArrayList<>();
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(Paths.get("shared_files"))) {
            for (Path p : stream) {
                if (Files.isRegularFile(p) && !p.getFileName().toString().endsWith(".meta")) {
                    out.add(p.getFileName().toString());
                }
            }
        }
        return out;
    }

    public static void sendEncrypted(PeerConn pc, String payload) throws Exception {
        byte[][] enc = Crypto.encrypt(pc.key, payload.getBytes());
        String line = "DATA|" +
                Base64.getEncoder().encodeToString(enc[0]) + "|" +
                Base64.getEncoder().encodeToString(enc[1]) + "\n";
        pc.writer.write(line);
        pc.writer.flush();
    }

    public static byte[] sha256(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        return md.digest(data);
    }

    public static Path metadataPath(String filename) {
        return Paths.get("downloads", filename + ".meta");
    }

    public static void saveMetadata(String filename, String originName, byte[] originPub, byte[] hash, byte[] sig) throws Exception {
        String content = originName + "|" +
                Base64.getEncoder().encodeToString(originPub) + "|" +
                Base64.getEncoder().encodeToString(hash) + "|" +
                Base64.getEncoder().encodeToString(sig);
        Files.writeString(metadataPath(filename), content);
    }

    public static Object[] loadMetadata(String filename) throws Exception {
        String content = Files.readString(metadataPath(filename)).trim();
        String[] parts = content.split("\\|");
        if (parts.length != 4) throw new IOException("invalid metadata");

        String originName = parts[0];
        byte[] originPub = Base64.getDecoder().decode(parts[1]);
        byte[] hash = Base64.getDecoder().decode(parts[2]);
        byte[] sig = Base64.getDecoder().decode(parts[3]);

        return new Object[]{originName, originPub, hash, sig};
    }

    public static byte[] loadDownload(String filename) throws Exception {
        String content = Files.readString(Paths.get("downloads", filename)).trim();
        String[] parts = content.split("\\|");
        if (parts.length != 2) throw new IOException("invalid stored file");

        byte[] nonce = Base64.getDecoder().decode(parts[0]);
        byte[] ciphertext = Base64.getDecoder().decode(parts[1]);

        return Crypto.decrypt(sha256("local-secret-key".getBytes()), nonce, ciphertext);
    }

    public static void processPayload(Main.App app, PeerConn pc, String payload) throws Exception {
        String[] parts = payload.split("\\|");
        String cmd = parts[0];

        if (cmd.equals("PING")) {
            System.out.println("[" + pc.name + "] PING received");
        } else if (cmd.equals("LIST_REQ")) {
            String joined = String.join(",", listSharedFiles());
            sendEncrypted(pc, "LIST_RES|" + joined);
        } else if (cmd.equals("LIST_RES")) {
            String files = parts.length > 1 ? parts[1] : "";
            System.out.println("[" + pc.name + "] shared files: " + (files.isEmpty() ? "(none)" : files));
        } else if (cmd.equals("GET_REQ")) {
            if (parts.length < 2) {
                sendEncrypted(pc, "ERROR|missing filename");
                return;
            }

            String filename = Paths.get(parts[1]).getFileName().toString();

            // ask user
            System.out.print("[" + pc.name + "] wants file '" + filename + "'. Accept? (y/n): ");

            BufferedReader console = new BufferedReader(new InputStreamReader(System.in));
            String resp = console.readLine();

            if (!resp.equalsIgnoreCase("y")) {
                sendEncrypted(pc, "ERROR|request denied");
                return;
            }

            Path sharedPath = Paths.get("shared_files", filename);
            Path downloadPath = Paths.get("downloads", filename);

            byte[] data;
            byte[] hash;
            byte[] sig;
            byte[] originPub;
            String originName;

            if (Files.exists(sharedPath)) {
                data = Files.readAllBytes(sharedPath);

                Path metaPath = Paths.get("shared_files", filename + ".meta");
                if (Files.exists(metaPath)) {
                    String content = Files.readString(metaPath).trim();
                    String[] meta = content.split("\\|");
                    originName = meta[0];
                    originPub = Base64.getDecoder().decode(meta[1]);
                    hash = Base64.getDecoder().decode(meta[2]);
                    sig = Base64.getDecoder().decode(meta[3]);
                } else {
                    hash = sha256(data);
                    sig = Handshake.sign(app.identity.priv, hash);
                    originPub = app.identity.pub.getEncoded();
                    originName = app.identity.name;
                }
            } else if (Files.exists(downloadPath)) {
                data = loadDownload(filename);
                Object[] meta = loadMetadata(filename);
                originName = (String) meta[0];
                originPub = (byte[]) meta[1];
                hash = (byte[]) meta[2];
                sig = (byte[]) meta[3];
            } else {
                sendEncrypted(pc, "ERROR|file not found");
                return;
            }

            String msg = "GET_RES|" + filename + "|" +
                    Base64.getEncoder().encodeToString(data) + "|" +
                    Base64.getEncoder().encodeToString(hash) + "|" +
                    Base64.getEncoder().encodeToString(sig) + "|" +
                    Base64.getEncoder().encodeToString(originPub) + "|" +
                    originName;
            sendEncrypted(pc, msg);
        } else if (cmd.equals("GET_RES")) {
            if (parts.length != 7) {
                System.out.println("[" + pc.name + "] malformed GET_RES");
                return;
            }
            String filename = parts[1];
            byte[] data = Base64.getDecoder().decode(parts[2]);
            byte[] hash = Base64.getDecoder().decode(parts[3]);
            byte[] sig = Base64.getDecoder().decode(parts[4]);

            byte[] actual = sha256(data);
            if (!Arrays.equals(actual, hash)) {
                System.out.println("[" + pc.name + "] hash mismatch for " + filename);
                return;
            }

            byte[] originPubRaw = Base64.getDecoder().decode(parts[5]);
            String originName = parts[6];

            Ed25519PublicKeyParameters originPub = new Ed25519PublicKeyParameters(originPubRaw, 0);

            if (!Handshake.verify(originPub, hash, sig)) {
                System.out.println("[" + pc.name + "] signature verification FAILED for "
                        + filename + " (original: " + originName + ")");
                return;
            }
            byte[][] enc = Crypto.encrypt(sha256("local-secret-key".getBytes()), data);
            String encryptedPayload = Base64.getEncoder().encodeToString(enc[0]) + "|" +
                    Base64.getEncoder().encodeToString(enc[1]);
            Files.writeString(Paths.get("downloads", filename), encryptedPayload);
            saveMetadata(filename, originName, originPubRaw, hash, sig);
            System.out.println("[" + pc.name + "] downloaded and verified " + filename + " (original: " + originName + ")");

        } else if (cmd.equals("KEY_UPDATE")) {
            if (parts.length < 2) {
                System.out.println("[" + pc.name + "] malformed KEY_UPDATE");
                return;
            }

            pc.remotePub = Base64.getDecoder().decode(parts[1]);
            System.out.println("[" + pc.name + "] updated public key");

        } else if (cmd.equals("ERROR")) {
            String msg = parts.length > 1 ? parts[1] : "unknown";
            System.out.println("[" + pc.name + "] ERROR: " + msg);
        }
    }
}
